### DISCUSSION TITLE
Illegal type Nothing

### DISCUSSION THREAD

**Discussion by garneja** [ClickHouse/ClickHouse#28606] [General]
How can I make this not try to evaluate an empty array as "Nothing"?

WITH
    [] AS A,
    [] AS M
SELECT arrayExists((a, m) -> bitAnd(a, m), A, M)

Received exception from server (version 21.3.3):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal types Nothing and Nothing of arguments of function bitAnd: While processing arrayExists((a, m) -> bitAnd(a, m), [] AS A, [] AS \
M).

**Comment by garneja** [comment]
This doesn't work either:

WITH
    [] AS A,
    [] AS M
SELECT if(notEmpty(A), arrayExists((a, m) -> bitAnd(a, m), A, M), 0)

Received exception from server (version 21.3.3):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Illegal types Nothing and Nothing of arguments of function bitAnd: While processing if(notEmpty([] AS A), arrayExists((a, m) -> bitAnd(a, m), A, [] AS M), 0).

**Comment by tavplubix** [comment]
Functions need to know argument types, but ClickHouse cannot deduce a type of empty array. You should specify some type explicitly, for example:
```
:) WITH []::Array(UInt8) AS A, []::Array(UInt8) AS M SELECT arrayExists((a, m) -> bitAnd(a, m), A, M)
```
or
```
:) WITH emptyArrayUInt8() AS A, emptyArrayUInt8() AS M SELECT arrayExists((a, m) -> bitAnd(a, m), A, M)
```
or
```
:) WITH CAST([], 'Array(UInt8)') AS A, CAST([] AS Array(UInt8)) AS M SELECT arrayExists((a, m) -> bitAnd(a, m), A, M)
```
